crates/handlers/src/compat/login.rs

```diff
@@ -368,12 +371,53 @@ pub(crate) async fn post(
         None
     };
 
+    // Ideally, we'd keep the lock whilst we actually create the device, but we
+    // really want to stop holding the transaction while we talk to the
+    // homeserver.
+    //
+    // In practice, this is fine, because:
+    // - the session exists after we commited the transaction, so a sync job won't
+    //   try to delete it
+    // - we've acquired a lock on the user before creating the session, meaning
+    //   we've made sure that sync jobs finished before we create the new session
+    // - we're in the read-commited isolation level, which means the sync will see
+    //   what we've committed and won't try to delete the session once we release
+    //   the lock
     repo.save().await?;
 
     activity_tracker
         .record_compat_session(&clock, &session)
         .await;
 
+    // This session will have for sure the device on it, both methods create a
+    // device
+    let Some(device) = &session.device else {
+        unreachable!()
+    };
+
+    // Now we can create the device on the homeserver, without holding the
+    // transaction
+    if let Err(err) = homeserver
+        .create_device(&user_id, device.as_str(), session.human_name.as_deref())
+        .await
+    {
+        // Something went wrong, let's end this session and schedule a device sync
+        let mut repo = repository_factory.create().await?;
+        let session = repo.compat_session().finish(&clock, session).await?;
```

Comment:
There is one slight difference in behaviour now: if the device creation failed on Synapse, before we'd just rollback the transaction, meaning that the client could retry the request with the same login token and it would work. This isn't the case anymore

It may also end another session, since we now end sessions with the same device ID

To be fair, this request was never really idempotent, so if the client lost the response it would have had the same effect